controllers/ocirepository_controller.go

```diff
@@ -942,3 +971,64 @@ func (r *OCIRepositoryReconciler) notify(ctx context.Context,
 		}
 	}
 }
+
+// notify emits notification related to the reconciliation.
+func (r *OCIRepositoryReconciler) verify(ctx context.Context, obj *sourcev1.OCIRepository, url string) (bool, error) {
+	// get the public keys from the given secret
+	secretRef := obj.Spec.Verify.SecretRef
+
+	ref, err := name.ParseReference(url)
+	if err != nil {
+		return false, err
+	}
+
+	if secretRef != nil {
+		certSecretName := types.NamespacedName{
+			Namespace: obj.Namespace,
+			Name:      secretRef.Name,
+		}
+
+		var pubSecret corev1.Secret
+		if err := r.Get(ctx, certSecretName, &pubSecret); err != nil {
+			return false, err
+		}
+
+		// traverse all public keys and try to verify the signature
+		// this is brute-force approach, but it is ok for now
+		for _, data := range pubSecret.Data {
+			pubRaw, err := base64.StdEncoding.DecodeString(string(data))
+			if err != nil {
+				return false, err
+			}
+
+			verifier, err := signature.New(signature.WithPublicKey(pubRaw))
+			if err != nil {
+				return false, err
+			}
+
+			signatures, _, err := verifier.VerifyImageSignatures(ctx, ref)
+			if err != nil {
+				return false, err
+			}
+
+			if len(signatures) > 0 {
+				return true, nil
+			}
+		}
+	} else {
+		verifier, err := signature.New()
+		if err != nil {
+			return false, err
+		}
+
+		signatures, _, err := verifier.VerifyImageSignatures(ctx, ref)
+		if err != nil {
+			return false, err
```

Comment:
This should at least state that the keyless verification failed.

Reply:
thank you, makes sense.